src: move Environment::context out of strong properties

joyeecheung opened the pull request:
Rename `ENVIRONMENT_STRONG_PERSISTENT_PROPERTIES` to
`ENVIRONMENT_STRONG_PERSISTENT_DATA`, and move `context`
out of the list, so that the data can be iterated separately.

<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

bnoordhuis commented on src/env.h:
```diff
@@ -340,11 +340,10 @@ constexpr size_t kFsStatsBufferLength = kFsStatsFieldsNumber * 2;
   V(zero_return_string, "ZERO_RETURN")
 
-#define ENVIRONMENT_STRONG_PERSISTENT_PROPERTIES(V)                            \
+#define ENVIRONMENT_STRONG_PERSISTENT_DATA(V)                                  \
```
You could merge this with `ENVIRONMENT_STRONG_PERSISTENT_VALUES` - `v8::Value` is a subclass of `v8::Data` too and everything in `ENVIRONMENT_STRONG_PERSISTENT_VALUES` is a subclass of `v8::Value`.

That said, since everything in `ENVIRONMENT_STRONG_PERSISTENT_DATA` is a subclass of `v8::Template` now, you should name it something like `PER_ISOLATE_TEMPLATE_PROPERTIES` and make them per-isolate.

joyeecheung replied:
Good point about making these `v8::Template`, though conceptually I think these are per-Environment, to add these into the snapshot I'll also need to register the external references for them, which is WIP, and the reason that these can be excluded from the minimal snapshot that we have (which is captured before Environment creation) is that none of these contains external references prior to Environment creation.

bnoordhuis commented on src/env-inl.h:
```diff
@@ -1106,10 +1106,13 @@ void AsyncRequest::set_stopped(bool flag) {
 
+  inline v8::Local<v8::Context> Environment::context() const {
+    return PersistentToLocal::Strong(context_);
+  }
```
Style: don't indent here.

joyeecheung commented:
Rebased and renamed `ENVIRONMENT_STRONG_PERSISTENT_DATA` to `ENVIRONMENT_STRONG_PERSISTENT_TEMPLATE` - we'll need separate routines to deserialize those for workers so I think it's best to put them in the environment for now.

joyeecheung commented:
> LGTM but can you add a TODO about changing them to per-isolate properties? Cheers.

@bnoordhuis What's the difference between per-isolate and per-Environment? So far in my brain, per-isolate stuff are more primitive and can be initialized without an Environment (i.e. before bootstrap), whereas per-Environment stuff are more sophisticated and can only be initialized after there is an associated Environment. By that standard the templates should be per-Environment because they are otherwise all empty.

addaleax commented:
I don’t have a strong opinion on per-Isolate vs per-Environment, but I don’t think it’s a bad idea to move Context-independent values to per-Isolate storage in general.

I think practical implications would only be there if we have multiple Environments per Isolate, which should be rare enough – 1. reducing the number of templates we create and 2. enabling the use of native objects from Environment A in Environment B… that doesn’t seem like an inherently bad thing.

joyeecheung commented:
@addaleax @bnoordhuis I took a look at the templates and I think so far most of them are per-Environment - for example most of the function templates have the Environment itself attached as callback data (for `Environment::GetCurrent()` to work), it does not seem to make sense to make these per-Isolate if there are multiple Environments operating on one isolate..

joyeecheung commented:
Landed in Landed in 6f02f15. We'll see if these templates can be made per-Isolate afterwards..
